fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -195,22 +187,33 @@ def reduction_fn() -> None:
                     if sz > self.buffer.numel():
                         # reduce big params directly
                         assert param_rank is not None
-                        reduce_params([param], cast(int, param_rank))
+                        reduce_grads([param], cast(int, param_rank))
                     else:
                         # smaller params are packed together from the same device
                         # and same rank.
                         if offset + sz > self.buffer.numel() or (
                             last_param_rank is not None and last_param_rank != param_rank
                         ):
                             assert last_param_rank is not None
-                            reduce_params(buffered_params, cast(int, last_param_rank))
+                            reduce_grads(buffered_params, cast(int, last_param_rank))
                             offset = 0
                             buffered_params.clear()
                         buffered_params.append(cast(Parameter, param))
                         offset += sz
 
                 if len(buffered_params) > 0:
                     assert param_rank is not None
-                    reduce_params(buffered_params, cast(int, param_rank))
+                    reduce_grads(buffered_params, cast(int, param_rank))
 
         reduction_fn()
+
+    def _sync_buffers(self) -> None:
+        _ = list(
+            map(
+                lambda x: x.wait(),
+                map(
+                    lambda x: dist.broadcast(x, self.authoritative_rank, self.process_group, async_op=True),
```

Comment:
does it make sense to combine some buffers and reduce the total number of broadcast, like what we do above for reduce_grad?

Reply:
yes, I think it would be an interesting option, I have an old branch doing that which I should revive. There's a trade-off though because there are copies in that case, in practice since the broadcast became async I saw a big speed bump and wondered whether this bucketing strategy is still to be followed ? I'll add a TODO, good idea